src/envoy/http/service_control/client_cache_test.cc

```diff
@@ -0,0 +1,179 @@
+namespace {
+
+using ::espv2::api_proxy::service_control::CheckResponseInfo;
+using ::google::api::envoy::http::service_control::FilterConfig;
+using ::google::api::envoy::http::service_control::Service;
+using ::google::api::servicecontrol::v1::CheckError;
+using ::google::protobuf::util::Status;
+using ::google::protobuf::util::error::Code;
+using ::testing::NiceMock;
+
+class ClientCacheCheckResponseTest : public ::testing::Test {
+ protected:
+  ClientCacheCheckResponseTest() : stats_base_("test", context_.scope_) {
+    token_fn_ = []() { return Envoy::EMPTY_STRING; };
+  }
+
+  void SetUp() override {
+    cache_ = std::make_unique<ClientCache>(
+        service_config_, filter_config_, stats_base_.stats(), cm_, time_source_,
+        dispatcher_, token_fn_, token_fn_);
+  }
+
+  void CheckAndReset(Envoy::Stats::Counter& counter, const int expected_value) {
+    EXPECT_EQ(counter.value(), expected_value);
+    counter.reset();
+  }
+
+  void TearDown() override {
+    CheckAndReset(stats_base_.stats().filter_.allowed_control_plane_fault_, 0);
+    CheckAndReset(stats_base_.stats().filter_.denied_control_plane_fault_, 0);
+  }
+
+  // Helpers for SetUp.
+  Service service_config_;
+  FilterConfig filter_config_;
+  NiceMock<Envoy::Upstream::MockClusterManager> cm_;
+  NiceMock<Envoy::Event::MockDispatcher> dispatcher_;
+  NiceMock<Envoy::MockTimeSystem> time_source_;
+  NiceMock<Envoy::Server::Configuration::MockFactoryContext> context_;
+  ServiceControlFilterStatBase stats_base_;
+  std::function<const std::string&()> token_fn_;
+
+  // Class under test.
+  std::unique_ptr<ClientCache> cache_;
+};
+
+TEST_F(ClientCacheCheckResponseTest, Http5xxAllowed) {
+  CheckDoneFunc on_done = [](const Status& status, const CheckResponseInfo&) {
+    EXPECT_EQ(status.code(), Code::OK);
+  };
+
+  CheckResponse* resp = new CheckResponse();
+  const Status http_status(Code::UNAVAILABLE, "");
+  cache_->handleCheckResponse(http_status, resp, on_done);
+
+  CheckAndReset(stats_base_.stats().filter_.allowed_control_plane_fault_, 1);
+}
+
+TEST_F(ClientCacheCheckResponseTest, Http4xxTranslatedAndBlocked) {
+  CheckDoneFunc on_done = [](const Status& status, const CheckResponseInfo&) {
+    EXPECT_EQ(status.code(), Code::INTERNAL);
+  };
+
+  CheckResponse* resp = new CheckResponse();
+  const Status http_status(Code::PERMISSION_DENIED, "");
+  cache_->handleCheckResponse(http_status, resp, on_done);
+}
+
+TEST_F(ClientCacheCheckResponseTest, Sc5xxAllowed) {
+  CheckDoneFunc on_done = [](const Status& status, const CheckResponseInfo&) {
+    EXPECT_EQ(status.code(), Code::OK);
+  };
+
+  CheckResponse* resp = new CheckResponse();
+  CheckError* check_error = resp->mutable_check_errors()->Add();
+  check_error->set_code(CheckError::NAMESPACE_LOOKUP_UNAVAILABLE);
+
+  const Status http_status(Code::OK, "");
+  cache_->handleCheckResponse(http_status, resp, on_done);
+
+  CheckAndReset(stats_base_.stats().filter_.allowed_control_plane_fault_, 1);
```

Comment:
Do we need to add integration tests for this change? I feel stats is a trivial component of envoy that we can expect to work, and we have unit tests to ensure they are incremented. We already have integration tests for some stats, I don't see the point of integration testing every single stat we add.

What do you think @qiwzhang, @TAOXUY?

Reply:
agree